File: go/quicksort/quicksort.go
package quicksort

func quicksort(a []int) []int {
	if len(a) < 2 {
		return a
	} else {
		pivot := a[0]
		less := findArrayLessEqualThanPivot(a[1:], pivot)
		greater := findArrayGreaterThanPivot(a[1:], pivot)

		return composeArray(composeArray(quicksort(less), []int{pivot}), quicksort(greater))
	}
}

func findArrayLessEqualThanPivot(a []int, pivot int) []int {
	r := make([]int, 0)

	for _, v := range a {
		if v <= pivot {
			r = append(r, v)
		}
	}

	return r
}

func findArrayGreaterThanPivot(a []int, pivot int) []int {
	r := make([]int, 0)

	for _, v := range a {
		if v > pivot {
			r = append(r, v)
		}
	}

	return r
}

func composeArray(a []int, b []int) []int {
	if len(a) == 0 {
		return b
	} else if len(b) == 0 {
		return a
	} else {
		return append(a, b...)
	}
}

File: SwiftAlgorithms/SwiftAlgorithms/QuickSort/QuickSort.swift
//
//  QuickSort.swift
//  SwiftAlgorithms
//
//  Created by ke on 2024/7/2.
//

import Foundation

struct QuickSort {
    static func quickSort(_ a: inout [Int]) -> [Int] {
        if a.count < 2 {
            return a
        } else {
            let index = a.count / 2
            let pivot = a[index]
            
            a.remove(at: index)
            var less = findArrayLessEqualThanPivot(a, pivot)
            var greater = findArrayGreaterThanPivot(a, pivot)
            
            return quickSort(&less) + [pivot] + quickSort(&greater)
        }
        
        
        
    }
    
    static func findArrayLessEqualThanPivot(_ a: [Int], _ pivot: Int) -> [Int] {
        var r: [Int] = []
        
        for v in a {
            if v <= pivot {
                r.append(v)
            }
        }
        
        return r
    }
    
    static func findArrayGreaterThanPivot(_ a: [Int], _ pivot: Int) -> [Int] {
        var r: [Int] = []
        
        for v in a {
            if v > pivot {
                r.append(v)
            }
        }
        
        return r
    }
}
